### DISCUSSION TITLE
query parameters with dashes instead of underscores

### DISCUSSION THREAD

**Discussion by Subaku** [fastapi/fastapi#7732] [Questions] [answered]
Simple question that I can't seem to find an answer to. The standard for python variables is to use underscores but when it comes to url part or query parameters it's often you'll see dashes instead. This is what I use everywhere else. In PyCharm I'm getting an error declaring my view parameters as such: "Illegal target for variable annotation". 

Realize this isn't FastAPI's fault but wasn't sure where to ask tbh. Any way to use dashes?

**Comment by dmontagu** [comment, marked as answer]
You can handle query parameters like this:

```python
from fastapi import FastAPI, Query
from starlette.testclient import TestClient

app = FastAPI()


@app.get("/")
def get_value(query_param: str = Query(..., alias="query-param")) -> str:
    return query_param


assert TestClient(app).get("/?query-param=1").json() == "1"
```

I'm looking at path parameters but it looks like the same trick won't work.

**Comment by dmontagu** [comment]
If you look at `starlette.routing.PARAM_REGEX`, it looks like hyphens are explicitly disallowed in path parameters:

```python
# Match parameters in URL paths, eg. '{param}', and '{param:int}'
PARAM_REGEX = re.compile("{([a-zA-Z_][a-zA-Z0-9_]*)(:[a-zA-Z_][a-zA-Z0-9_]*)?}")
```
In particular, they must only consist of alphanumeric characters or `_`.

I suspect if this restriction was weakened to allow hyphens the same trick as for query parameters above would work to add hyphens to path parameters.

If you care about this, I would recommend creating an issue in the starlette repo.

**Comment by Subaku** [comment]
I appreciate your help! That's certainly interesting. I'll go poke that repo and see what's up :)

Thanks again.

**Comment by tiangolo** [comment]
Thanks for the help here @dmontagu ! :cake: :bowing_man: 

Thanks @Subaku for reporting back and closing the issue :+1:

**Comment by HansBrende** [comment]
@Subaku @tiangolo @dmontagu I found a little hack to get dashed query parameters for all one's routes without having to specify a hundred different aliases:

```python
def DashingQuery(default: Any, *, convert_underscores=True, **kwargs) -> Any:
    query = Query(default, **kwargs)
    query.convert_underscores = convert_underscores
    return query
```

Now we can do:
```python
@app.get("/")
def get_value(query_param: str = DashingQuery(...)) -> str:
    return query_param
```

Couldn't this be supported out of the box with just a 1 or 2-line tweak since the functionality is already present for `Header`?

Here's a 1-liner in `fastapi.params.Query.__init__` that would do the trick:
```python
self.convert_underscores = extra.pop('convert_underscores', False)
```
